### PR TITLE
backupccl: correctly manage temp objects during a full cluster restore

### LINKED ISSUE #56512: bulkio: backups with temporary tables leave descriptors in weird state
When performing a database/cluster restore with temporary tables, a database's descriptors may reach an unexpected state.

In 20.1, this can be achieved with a database restore. In 20.2, this can only be achieved with a cluster restore (of a cluster with temporary tables).

After restoring a backup that contains temp tables, performing a `SHOW TABLES` in the cluster leads to:
```
root@localhost:26257/defaultdb> show tables;
ERROR: internal error: schema id 52 not found
SQLSTATE: XX000
DETAIL: stack trace:
/go/src/github.com/cockroachdb/cockroach/pkg/sql/information_schema.go:1940: forEachTableDescWithTableLookupInternal()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/information_schema.go:1812: forEachTableDescAllWithTableLookup()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/information_schema.go:1788: forEachTableDescAll()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:404: func53()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:492: 1()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:115: func2()
/usr/local/go/src/runtime/asm_amd64.s:1357: goexit()
```

ID 52 refrences the pg_temp schema:
```
> select * from system.namespace;
  parentID | parentSchemaID |              name               | id
-----------+----------------+---------------------------------+-----
[...]
        50 |              0 | public                          | 29
        50 |             52 | a_temp_table                    | 53
        51 |              0 | public                          | 29
        54 |             29 | t                               | 55
```

One solution here could be to not restore any temporary tables back during cluster restore. However, in this case we need to consider possible jobs that could be referencing these descriptors. When a session ends, the temporary tables are dropped and we create a job (including a GC job, which references this descriptor). If cluster restore chooses not to restore temporary tables, then it should also choose to not restore jobs associated with their tables.

### PR DESCRIPTION
**Pull request by adityamaru**
Discussion about the design at #56512.

Fixes: #56512

Release note: None

### CODE REVIEW

File: pkg/ccl/backupccl/restore_planning.go [hunk 1]
```diff
@@ -332,6 +370,64 @@ func allocateDescriptorRewrites(
 				descriptorRewrites[typ.GetID()] = &jobspb.RestoreDetails_DescriptorRewrite{ParentID: tempSysDBID}
 			}
 		}
+
+		// When restoring a temporary object, the parent schema which the descriptor
```
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_planning.go:374]: let me know if this code is too comment heavy!

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56956)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @pbardea)

___
*[pkg/ccl/backupccl/backup_test.go, line 126 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56956#-MMaSkkGk0wv6ozld_D3-r2-126:-MMpISR11_gKZY_cSyZl:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/0acc7cc2b2dc037181bf16181564a560a220916b/pkg/ccl/backupccl/backup_test.go#L126)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

it may be safer to error out here if durations are invalid, otherwise it is silent.
</blockquote></details>

done.
___
*[pkg/ccl/backupccl/backup_test.go, line 6788 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56956#-MMaSkkGk0wv6ozld_D3-r2-6788:-MMpIYVBBFfJUO19aNW4:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/0acc7cc2b2dc037181bf16181564a560a220916b/pkg/ccl/backupccl/backup_test.go#L6788)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

remove?
</blockquote></details>

done.
___
*[pkg/ccl/backupccl/restore_planning.go, line 426 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56956#-MMaSkkGk0wv6ozld_D6-r2-426:-MMpKPyC1-JagxydBnEb:b658600) ([raw file](https://github.com/cockroachdb/cockroach/blob/0acc7cc2b2dc037181bf16181564a560a220916b/pkg/ccl/backupccl/restore_planning.go#L426)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

can this collide? the IDs do not appear global. is there a test for it?
</blockquote></details>

Restore planning happens only on the coordinator node (as opposed to job execution) so there should be no collisions when processing the temp objects and assigning them fake temp schemas.
The tests I added do require multiple fake schemas to be created on restore.

**Review by pbardea — APPROVED** [review]
LGTM with just a few nits

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56956#-:-MMpXBOr0yXlim4LkSZN:bjphn42)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @otan, and @pbardea)

___
*[pkg/ccl/backupccl/backup_test.go, line 229 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56956#-MMpS95K6mMsDf677r2D:-MMpS95LArraBE_pVPDH:b-auoe7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/99902bb30699a14b282aca78386dbb751d65e250/pkg/ccl/backupccl/backup_test.go#L229)):*
> ```Go
> 					allowImplicitAccess = true
> 				}
> 				if d.HasArg("temp-cleanup-freq") {
> ```

It might be more clear to instead set the `sql.temp_object_cleaner.cleanup_interval` cluster setting in the test, rather than plumb it through the server startup?
___
*[pkg/ccl/backupccl/backup_test.go, line 6754 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56956#-MMpWF-_23Ev3u0Kmuc3:-MMpWF-_23Ev3u0Kmuc4:b-ym8fr5) ([raw file](https://github.com/cockroachdb/cockroach/blob/99902bb30699a14b282aca78386dbb751d65e250/pkg/ccl/backupccl/backup_test.go#L6754)):*
> ```Go
> 	// have different pg_temp schemas but will be created in the same defaultdb.
> 	comment := "never see this"
> 	for _, dbID := range []int{0, 1} {
> ```

nit: can we rename `dbID` and `db` here to `conn` to avoid confusion that the temp tables are created in the same DB?
___
*[pkg/ccl/backupccl/backup_test.go, line 6768 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56956#-MMpWckJD1men5pzusqM:-MMpWckJD1men5pzusqN:be5vbxl) ([raw file](https://github.com/cockroachdb/cockroach/blob/99902bb30699a14b282aca78386dbb751d65e250/pkg/ccl/backupccl/backup_test.go#L6768)):*
> ```Go
> 	diffDB := sqlutils.MakeSQLRunner(diffDBConn)
> 	diffDB.Exec(t, `SET experimental_enable_temp_tables=true`)
> 	diffDB.Exec(t, `CREATE DATABASE t1`)
> ```

nit: can we call the databases something like `d1` and `d2`?
___
*[pkg/ccl/backupccl/backup_test.go, line 6797 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56956#-MMpWyoR4LICNO4PQMac:-MMpWyoR4LICNO4PQMad:bb3uc43) ([raw file](https://github.com/cockroachdb/cockroach/blob/99902bb30699a14b282aca78386dbb751d65e250/pkg/ccl/backupccl/backup_test.go#L6797)):*
> ```Go
> 	defer cleanupRestore()
> 	sqlDBRestore.Exec(t, `RESTORE FROM 'nodelocal://0/full_cluster_backup'`)
> ```

Can we add an extra sanity check that `SHOW TABLES` works as expected in one of the dbs before the reconciliation job kicks in?

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56956)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @pbardea)

___
*[pkg/ccl/backupccl/backup_test.go, line 229 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56956#-MMpS95K6mMsDf677r2D:-MMq-WEV8VMFJkZjCMWg:b6m2yox) ([raw file](https://github.com/cockroachdb/cockroach/blob/99902bb30699a14b282aca78386dbb751d65e250/pkg/ccl/backupccl/backup_test.go#L229)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

It might be more clear to instead set the `sql.temp_object_cleaner.cleanup_interval` cluster setting in the test, rather than plumb it through the server startup?
</blockquote></details>

So I did try that but I think the issue is that the reconciliation job reads the cluster setting immediately after running the first time - https://github.com/cockroachdb/cockroach/blob/be5a871b99151b17a93981666f246ae15699cc49/pkg/sql/temporary_schema.go#L610

So if this setting is not set at test startup, it will only read the updated cluster setting after the second run, which will by default occur 30mins after the first. So to make the test deterministic the server needs to have the cluster setting overriden on startup.
___
*[pkg/ccl/backupccl/backup_test.go, line 6754 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56956#-MMpWF-_23Ev3u0Kmuc3:-MMq1RFf1yyHtBC7bnJx:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/99902bb30699a14b282aca78386dbb751d65e250/pkg/ccl/backupccl/backup_test.go#L6754)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

nit: can we rename `dbID` and `db` here to `conn` to avoid confusion that the temp tables are created in the same DB?
</blockquote></details>

done.
___
*[pkg/ccl/backupccl/backup_test.go, line 6768 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56956#-MMpWckJD1men5pzusqM:-MMqA-A66S1UCNff6reu:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/99902bb30699a14b282aca78386dbb751d65e250/pkg/ccl/backupccl/backup_test.go#L6768)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

nit: can we call the databases something like `d1` and `d2`?
</blockquote></details>

done.
___
*[pkg/ccl/backupccl/backup_test.go, line 6797 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56956#-MMpWyoR4LICNO4PQMac:-MMq9vO8BhvGM5k_FCa5:b-q2e8yy) ([raw file](https://github.com/cockroachdb/cockroach/blob/99902bb30699a14b282aca78386dbb751d65e250/pkg/ccl/backupccl/backup_test.go#L6797)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Can we add an extra sanity check that `SHOW TABLES` works as expected in one of the dbs before the reconciliation job kicks in?
</blockquote></details>

done. (they'll be empty due to the remapping)

**Comment by adityamaru** [comment]
TFTR!

bors r=otan,pbardea

**Pull request merged**